borsh/src/schema_helpers.rs

```diff
@@ -36,3 +36,208 @@ pub fn schema_container_of<T: BorshSchema>() -> BorshSchemaContainer {
+/// This is used by [`BorshSchema::max_serialized_size()`] to handle weird types
+/// such as `[[[(); u32::MAX]; u32::MAX]; u32::MAX]` which serialises to an
+/// empty string even though its number of elements overflows `usize`.
+fn is_zero_size(declaration: &str, defs: &BTreeMap<Declaration, Definition>) -> bool {
+    match defs.get(declaration).ok_or(declaration) {
+        Ok(Definition::Array { length, elements }) => {
+            *length == 0 || is_zero_size(elements.as_str(), defs)
+        }
+        Ok(Definition::Sequence { .. }) => false,
+        Ok(Definition::Tuple { elements }) => elements
+            .into_iter()
+            .all(|element| is_zero_size(element.as_str(), defs)),
+        Ok(Definition::Enum { .. }) => false,
+        Ok(Definition::Struct { fields }) => match fields {
+            Fields::NamedFields(fields) => fields
+                .into_iter()
+                .all(|(_, field)| is_zero_size(field.as_str(), defs)),
+            Fields::UnnamedFields(fields) => fields
+                .into_iter()
+                .all(|field| is_zero_size(field.as_str(), defs)),
+            Fields::Empty => true,
+        },
+
+        Err(dec) => dec == "nil",
+    }
+}
+
+/// Implementation of [`BorshSchema::max_serialized_size`].
+fn max_serialized_size_impl(
+    count: usize,
+    declaration: &str,
+    defs: &BTreeMap<Declaration, Definition>,
+) -> Option<usize> {
+    use core::convert::TryFrom;
+
+    /// Maximum number of elements in a vector or length of a string which can
+    /// be serialised.
+    ///
+    /// This is `u32::MAX` cast to `usize`.
+    const MAX_LEN: usize = 4294967295;
+
+    /// Calculates max serialised size of a tuple with given members.
+    fn tuple<'a>(
+        count: usize,
+        elements: impl core::iter::IntoIterator<Item = &'a Declaration>,
+        defs: &BTreeMap<Declaration, Definition>,
+    ) -> Option<usize> {
+        let mut sum: usize = 0;
+        for el in elements {
+            sum = sum.checked_add(max_serialized_size_impl(1, el, defs)?)?;
+        }
+        count.checked_mul(sum)
+    }
+
+    match defs.get(declaration).ok_or(declaration) {
+        Ok(Definition::Array { length, elements }) => {
+            // Aggregate `count` and `length` to a single number.  If this
+            // overflows, check if array’s element is zero-sized.
+            let count = usize::try_from(*length)
+                .ok()
+                .and_then(|len| len.checked_mul(count));
+            match count {
+                Some(0) => Some(0),
+                Some(count) => max_serialized_size_impl(count, elements, defs),
+                None if is_zero_size(elements, defs) => Some(0),
+                None => None,
+            }
+        }
+        Ok(Definition::Sequence { elements }) => {
+            // Assume that sequence has MAX_LEN elements since that’s the most
+            // it can have.
+            let sz = max_serialized_size_impl(MAX_LEN, elements, defs)?;
```

Comment:
Preallocating 4GB of memory for `String` or `Vec<u8>`, or `HashMap<u64, u32>` doesn't appear terribly useful, if one instead expects the type to casually contain 200 bytes, or 2MB.

Using a number of types like [BoundedVec](https://docs.rs/bounded-vec/latest/bounded_vec/) you mentioned instead might be better imo.
Thus there're 2 alternatives:

1. Add `max_length` to [Definition::Sequence](https://github.com/mina86/borsh-rs/blob/8ae9f6c580790548509b64b4dda80ef2a51438d7/borsh/src/schema.rs#L41) :
```rust
    Sequence { max_length: u32, elements: Declaration },
```
which will usually be populated with `u32::MAX`, and with something less for types such as `BoundedVec<T, L, U>`.
In this variant it can be made to work with statically bounded types with current helper function `max_serialized_size` after small adjustments.

2. To allow to compute max size for types, whose limit may be set at runtime:

```rust
    let mut rng = rand::thread_rng();
    let up: u64 = rng.next_u64() % 100 + 1;
    let up1: usize = up as usize;
    // error[E0435]: attempt to use a non-constant value in a constant
    let vec: BoundedVec<u8, 0, up1> = BoundedVec::from_vec(vec![20, 20, 20]).unwrap();
```
`pub fn max_serialized_size<T: BorshSchema>() -> Option<usize>` may be reverted back to be a 
`BorshSchema` method, but with `&self` parameter, unused in default implementation:

```rust
    fn max_serialized_size(&self) -> Option<usize> {
    ...
    }
```

@frol what do you think?

Side note: i've noticed that derived implementation of [`serde`](https://github.com/ergoplatform/bounded-vec/blob/develop/src/bounded_vec.rs#L10) for `BoundedVec<T, L, U>` doesn't attempt to check the imposed length bounds at all.

Reply:
For the second point, this would require changing schema so that rather than having a mapping from Declaration to Definition, there would be a mapping from Declaration to `&dyn BorshSchema` and then additional `fn definition(&self) -> Option<Definition>` method.  Or Definition would need to be changed to be a trait with `definition(&self) -> Option<Definition>;` and `max_serialized_size(&self) -> Option<usize>` methods.

However, problem with that approach is that this isn’t serialisable.  If the idea behind BorshSchema is to be able to read encoded data without having the definition of the types, it’s potentially pointless to rely on traits and such to get the max size information.

Maybe I’m overthinging this, but I’m imagining something like:

```rust
pub enum Definition {
    /// A sequence of elements of the same type.
    Sequence {
        /// How many bytes does the length prefix occupy.
        ///
        /// Zero means this is a fixed-length array or a custom encoding schema
        /// which doesn’t follow typical borsh conventions.  In the former case,
        /// `min_length` and `max_length` will be set to the same value.  In the
        /// latter case, it may be impossible to deserialise the data without
        /// additional knowledge of the declared type.
        length_width: u8,

        /// Minimum length of the sequence.  Usually `0` for dynamically-sized
        /// sequences.
        min_length: u64,

        /// Maximum length of the sequence.  Usually `2**length_width - 1`.
        max_length: u64,

        /// Type of the elements of the sequence.
        elements: Declaration,
    },

    /// A fixed-size tuple with the length known at the compile time and the
    /// elements of different types.
    Tuple {
        /// Types of elements of the tuple.
        elements: Vec<Declaration>,
    },

    /// A tagged union, a.k.a enum. Tagged-unions have variants with associated
    /// structures.
    Enum {
        /// Width of the discriminant tag.
        tag_width: u8,

        /// Possible variants of the enumeration.
        variants: Vec<(VariantName, Declaration)>,
    },

    /// A structure, structurally similar to a tuple.
    Struct {
        /// Named fields of the structure with their types.
        ///
        /// Tuple-like structures are modelled as a `Tuple`.
        fields: Vec<(FieldName, Declaration)>,
    },
}

```

With that
* `[u8; 10]` would be `Sequence { length_width: 0, min_length: 10, max_length: 10, elements: "u8" }`,
* `SmallVec<u8>` would be `Sequence { length_width: 1, min_length: 0, max_length: 255, elements: "u8" }`,
* `BoundVec<1, 10, u8>` would be `Sequence { length_width: 4, min_length: 1, max_length: 10, elements: "u8" }` and
* `VarInt<u32>` would be `Sequence { length_width: 0, min_length: 1, max_length: 5, elements: "u8" }`.

Reply:
`Sequence { length_width: u8, min_length: u64, max_length: u64, elements: Declaration }` doesn't look that bad, especially due to being able to accommodate `VarInt`. 

I would still not remove `Definition::Array`, though `Sequence { length_width: 0, min_length: 10, max_length: 10, elements: "u8" }` is fine to be understood too.

`tag_width` for `Enum` is redundant imo, as `Enum` is supposed to be human-readable, and if one has an `Enum` with over-than-u8-range variants, or over-than-u16-range variants, he/she probably has to factor something out, to make the definition of `Enum` human readable again.

Reply:
I like explicitness, so I like the proposed extended `Definition`